### ISSUE TITLE
Current MEADS implementation is incomplete

### ISSUE THREAD

**Issue by alexlyttle** [blackjax-devs/blackjax#781]
### Current behavior

The MEADS adaptation routine appears to be incomplete. Currently, cross-chain statistics are computed each iteration and used to update the kernel parameters for the entire chain. This is missing some aspects of Algorithm 3 from [Hoffman & Sountsov (2022)](https://proceedings.mlr.press/v151/hoffman22a/hoffman22a.pdf). Perhaps this was on purpose, in which case I would be interested to know why.

```python
meads = blackjax.meads_adaptation(logdensity_fn, num_chains)
```

### Desired behavior

[Hoffman & Sountsov (2022)](https://proceedings.mlr.press/v151/hoffman22a/hoffman22a.pdf) describe an algorithm where the chains are split into $K$ folds. Cross-chains statistics are computed within each fold and used to update the neighbouring fold each iteration (skipping the fold equal to the current iteration modulo $K$). It also describes a shuffling of all chains every $K$ steps. It appears the original author implemented the algorithm in [this notebook](https://github.com/tensorflow/probability/blob/main/discussion/meads/meads.ipynb). I have recently experimented with modifying the BlackJAX MEADS to reflect this for a project testing new MCMC adaptation algorithms.

I propose updating the existing implementation to include the $K$-folding and shuffling described in the paper. This would introduce a few more parameters to `blackjax.meads_adaptation` which could take default values from the paper.

```python
meads = blackjax.meads_adaptation(logdensity_fn, num_chains, num_folds=4, shuffle=True, step_size_multiplier=0.5, damping_slowdown=1.0)
```

The `step_size_multiplier` and `damping_slowdown` are hyper-parameters used in calculating the MEADS statistics.

**Comment by junpenglao** [comment]
@albcab thoughts?

**Comment by albcab** [comment]
Hi @alexlyttle,

you are right, the current adaptation routine is not exactly the one used in the paper. It was a long time ago, but I think at the end I wanted it to be an adaptation algorithm. MEADS is an adaptive algorithm (as in real adaptive MCMC, which you should never stop adapting and you are sure it will converge to your distribution). Instead, adaptation algorithms are meant to stop adapting and then fix a kernel that you use to generate your samples, a kernel that doesn't change, so you don't need to worry about changing hyperparameters ruining convergence results.

When I wrote the original code, adaptation algorithms would run for a specified number of steps and would return a kernel (GHMC in this case) to fix and use ([here](https://github.com/blackjax-devs/blackjax/pull/229) is the original code and discussion). Real MEADS would've returned $K$ kernels, which was an API problem. I also opened a pull request to make it more adaptive-ish [here](https://github.com/blackjax-devs/blackjax/pull/442), but it was not merged, can't remember why. 

Anyway, API has changed, adaptation algorithms have stepping functions and adaptive algorithms can more easily be implemented. So, by all means, go ahead and share/pull your complete adaptive MEADS!

**Comment by alexlyttle** [comment]
Thanks @albcab, super informative! My follow-up question would have been about how the paper presents MEADS as an adaptive algorithm rather than something where you'd need to freeze the kernel. This makes sense though, I can see your reasoning in the original PR.

I am happy to share my version and contribute if it's welcome.

Practically, should this be added as its own MCMC algorithm in BlackJAX? My understanding of the paper is that MEADS is to GHMC what NUTS is to HMC.

**Comment by albcab** [comment]
That is a good question, more of a code design/API question. You could either extend the current MEADS implementation in `adaptation/` or create a new one in `mcmc/`. As you said, MEADS is a bit more like NUTS than window adaptation, and there are both `hmc.py` and `nuts.py` in `mcmc/`. But NUTS is kind of special in bayesian computation because of how good it is and thus how widely it is used. Though I'm not sure if you can also adapt the step size as a real adaptive MCMC along with NUTS. If not, that would make MEADS more "complete" than NUTS as adaptive MCMC.

I'm not up to date with code design in the library, so maybe @junpenglao is better for deciding this? I'd lean towards making a new algorithm in `mcmc/` like NUTS.

**Comment by reubenharry** [comment]
@ColCarroll Note that the incoming [EMAUS](https://github.com/blackjax-devs/blackjax/pull/766) implementation might have some relevant code (see e.g. `run_eca`), although not sure it's relevant.

**Comment by albcab** [comment]
+1 what @ColCarroll said about having ECA as its own thing so it can be used with other algorithms.
